### PR TITLE
Add biased-decoding feature

### PR DESCRIPTION
**Pull request by scotfang**

### CODE REVIEW

**Comment by scotfang** [comment]
Hi, this PR contains all the code and tests needed for the new anti-flicker feature I've implemented.  Please take a look when you have a chance, no rush!

File: src/decoding.cc [hunk 1]
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
...
+      log_prob_beam_views[i] = &(log_prob_beam_view_storage[i]);
+    }
+    ops::Split(0, /*no_copy=*/true)(logits, logit_beam_views);
+    log_probs.resize_as(logits);
+    log_probs.reshape(logits.shape());
+    ops::Split(0, /*no_copy=*/true)(log_probs, log_prob_beam_views);
+
+    // Scalar's need to be allocated on CPUs.
+    StorageView scalar_discount(1 - prefix_bias_beta, Device::CPU);
+    // const dim_t cur_beam_size = is_expanded ? _beam_size : 1;
```
Comment (guillaumekln) [inline src/decoding.cc:150]: Remove this line?

File: src/decoding.cc [hunk 2]
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
...
+        ops::Log()(_spare_beam, log_prob_beam);
+      } else {
+        ops::LogSoftMax()(logit_beam, log_prob_beam);
+      }
+    }
+  }
 
-  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty)
+  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty, 
+                         const float prefix_bias_beta)
```
Comment (guillaumekln) [inline src/decoding.cc:185]: Minor formatting suggestion:

```suggestion
  BeamSearch::BeamSearch(const dim_t beam_size,
                         const float length_penalty,
                         const float coverage_penalty, 
                         const float prefix_bias_beta)
```

File: src/decoding.cc [hunk 3]
```diff
@@ -189,18 +267,34 @@ namespace ctranslate2 {
...
               topk_ids.to(device),
               state,
-              &logits,
+              &logits,  // output shape: (cur_batch_size*beam_size x 1 x vocab_size), if not expanded beam_size is 1
               (attention || _coverage_penalty != 0) ? &attention_step_device : nullptr);
-      ops::LogSoftMax()(logits, log_probs);
+
+      if (bias_towards_prefix) {
+        if (!biased_decoder) {
+          biased_decoder.reset(new BiasedDecoder());
```
Comment (guillaumekln) [inline src/decoding.cc:281]: We try to follow this pattern for unique_ptr:

```suggestion
          biased_decoder = std::make_unique<BiasedDecoder>();
```

File: tests/translator_test.cc [hunk 4]
```diff
@@ -338,6 +339,295 @@ TEST(TranslatorTest, TranslateBatchWithPrefixAndEmpty) {
   EXPECT_EQ(result[4].output(), (std::vector<std::string>{"a", "z", "z", "a"}));
 }
 
+TEST(TranslatorTest, TranslateBatchWithStronglyBiasedPrefixAndEmpty) {
```
Comment (guillaumekln) [inline tests/translator_test.cc:342]: Empty inputs are handled outside of the decoding code, so I don't think it is relevant to include empty inputs when testing this feature.

I suggest removing the empty input and renaming the test case.

File: tests/translator_test.cc [hunk 5]
```diff
@@ -338,6 +339,295 @@ TEST(TranslatorTest, TranslateBatchWithPrefixAndEmpty) {
...
+    {}};
+  const auto result = translator.translate_batch_with_prefix(input, prefix, options);
+  EXPECT_EQ(result[0].output(), (std::vector<std::string>{"a", "t", "s", "u", "m", "o", "n"}));
+  EXPECT_EQ(result[1].output(), (std::vector<std::string>{"a", "t", "z", "m", "o", "n"}));
+  EXPECT_EQ(result[2].output(), (std::vector<std::string>{"a", "t", "z", "o", "m", "o", "n"}));
+  EXPECT_TRUE(result[3].output().empty());
+  EXPECT_EQ(result[4].output(), (std::vector<std::string>{"a", "z", "z", "a"}));
+}
+
+TEST(TranslatorTest, TranslateBatchWithWeaklyBiasedPrefixAndEmpty) {
```
Comment (guillaumekln) [inline tests/translator_test.cc:369]: Same comment regarding the empty input.

File: src/decoding.cc [hunk 6]
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
...
+
+    if (_spare_beam.dtype() != dtype || _spare_beam.device() != device) {
+      _spare_beam = StorageView(device, dtype);
+    }
+
+    std::vector<StorageView> logit_beam_view_storage(num_beams, StorageView(device, dtype));
+    std::vector<StorageView*> logit_beam_views(num_beams);
+    std::vector<StorageView> log_prob_beam_view_storage(num_beams, StorageView(device, dtype));
+    std::vector<StorageView*> log_prob_beam_views(num_beams);
+    for (int i = 0; i < num_beams; ++i) {
```
Comment (guillaumekln) [inline src/decoding.cc:139]:
```suggestion
    for (dim_t i = 0; i < num_beams; ++i) {
```

File: src/decoding.cc [hunk 7]
```diff
@@ -170,6 +239,15 @@ namespace ctranslate2 {
       attention->resize(batch_size);
     }
 
+    // TODO(scotfang): Instead of initializing and empty vector, use unique_ptr.
```
Comment (guillaumekln) [inline src/decoding.cc:242]: I don't think initializing an empty vector will cause any performance issue, so you can probably remove this comment.

**Review by guillaumekln — CHANGES_REQUESTED** [review]

**Comment by guillaumekln** [comment]
Can you also update the relevant documentation?

* Add the feature in README.md -> Supported decoding options
* Add the new arguments in the Python API reference in docs/python.md
* (optional) Add an example at the end of docs/decoding.md

File: include/ctranslate2/translator.h [hunk 8]
```diff
@@ -30,6 +30,15 @@ namespace ctranslate2 {
...
     // Coverage value to apply during beam search.
     float coverage_penalty = 0;
+    // Biases decoding towards a given prefix, see https://arxiv.org/abs/1912.03393 --section 4.2
+    // Only activates biased-decoding when beta is in range (0, 1) and SearchStrategy is set to BeamSearch.
+    //
+    // Setting beta to 1.0 is equivalent to using a hard prefix, setting it to 0 is equivalent
+    // to unconstrained decoding.  The higher beta is, the stronger the bias towards the given prefix.
+    //
+    // If you translate with a non-empty prefix and beta <= 0, 
+    // then the prefix will be used as a hard-prefix rather than a soft, biased-prefix.
```
[the remark this answers is not in the document]
Comment (scotfang) [reply include/ctranslate2/translator.h:40]: Made docs more clear.

File: src/decoding.cc (hunk 1 above)
Comment (scotfang) [reply src/decoding.cc:150]: done

File: tests/translator_test.cc (hunk 4 above)
Comment (scotfang) [reply tests/translator_test.cc:342]: done

File: tests/translator_test.cc (hunk 5 above)
Comment (scotfang) [reply tests/translator_test.cc:369]: done

File: src/decoding.cc (hunk 7 above)
Comment (scotfang) [reply src/decoding.cc:244]: done

**Comment by scotfang** [comment]
Addressed issues brought up in review, re-requested review.  

Should I also try to resolve conflicts? Some of the changes from master look new to me.

**Review by guillaumekln — APPROVED** [review]
Looks good! I resolved the conflicts in `decoding.cc` and apply some minor reformatting to match the existing code.

Thanks for your contribution!

**Comment by scotfang** [comment]
Thank you so much for merging this work!

**Pull request merged**